CupertinoSlidingSegmentedControl

LongCatIsLooong opened the pull request:
## Description

Introduce `CupertinoSlidingSegmentedControl`, the iOS 13 `CupertinoSegmentedControl`.

![segmented](https://user-images.githubusercontent.com/31859944/66857941-27209100-ef3d-11e9-84da-b5b225abbf91.gif)


## Related Issues

Closes https://github.com/flutter/flutter/issues/33796

## Tests

- add packages/flutter/test/cupertino/sliding_segmented_control_test.dart

## TODOs
- `momentary` is missing: https://developer.apple.com/documentation/uikit/uisegmentedcontrol/1618586-ismomentary
- Currently only supports `equalWidth` layout.  `proportional` is missing.
- The thumb shadow is done exactly as the sketch file describes, but it looks a bit off when compared with the native component.
- The drag gesture recognizer has to be passed to the render object because it needs to know children's dimensions. We probably need a `CupertinoMultiChildGestureDetector` or something, as this kind of gesture is quite common on iOS.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

LongCatIsLooong committed:
remove CupertinoSystemColors

LongCatIsLooong committed:
wrong color

LongCatIsLooong committed:
update docs

LongCatIsLooong committed:
comments

LongCatIsLooong committed:
update comments

LongCatIsLooong committed:
update tests

LongCatIsLooong committed:
s/gray/grey/

LongCatIsLooong committed:
Merge branch 'colors' into segmented-control

LongCatIsLooong committed:
WIP

LongCatIsLooong committed:
WIP

LongCatIsLooong committed:
WIP

LongCatIsLooong committed:
Merge remote-tracking branch 'upstream/master' into segmented-control

LongCatIsLooong committed:
update

LongCatIsLooong committed:
shadow

LongCatIsLooong committed:
update

LongCatIsLooong committed:
update

LongCatIsLooong committed:
Merge remote-tracking branch 'upstream/master' into segmented-control

LongCatIsLooong committed:
update

LongCatIsLooong committed:
Pog

LongCatIsLooong committed:
use simulation instead of curve

LongCatIsLooong committed:
create new class for segmeneted control

LongCatIsLooong committed:
update

LongCatIsLooong committed:
Merge branch 'master' of github.com:flutter/flutter into segmented-control

LongCatIsLooong committed:
update

LongCatIsLooong committed:
Merge remote-tracking branch 'upstream/master' into segmented-control

LongCatIsLooong committed:
update

LongCatIsLooong committed:
update

LongCatIsLooong committed:
Merge branch 'master' of github.com:flutter/flutter into segmented-control

LongCatIsLooong committed:
WIP

LongCatIsLooong committed:
Merge remote-tracking branch 'upstream/master' into segmented-control

LongCatIsLooong committed:
add tests

LongCatIsLooong committed:
update tests

LongCatIsLooong committed:
Merge branch 'master' of github.com:flutter/flutter into segmented-control

LongCatIsLooong committed:
update

LongCatIsLooong committed:
update

LongCatIsLooong committed:
tests passed

LongCatIsLooong committed:
clean up

LongCatIsLooong committed:
Merge remote-tracking branch 'upstream/master' into segmented-control

LongCatIsLooong committed:
move drag to state

LongCatIsLooong committed:
remove import material

LongCatIsLooong committed:
interactive example

justinmc commented on packages/flutter/test/cupertino/sliding_segmented_control_test.dart:
```diff
@@ -0,0 +1,924 @@
+
+      await tester.tap(find.byIcon(const IconData(1)));
+      await tester.pump();
+      await tester.pump(const Duration(milliseconds: 500));
```
Does `pumpAndSettle` not work?

justinmc commented on packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:
```diff
@@ -0,0 +1,1037 @@
+/// See also:
+///
+///  * <https://developer.apple.com/design/human-interface-guidelines/ios/controls/segmented-controls/>
+class CupertinoSlidingSegmentedControl<T> extends StatefulWidget {
```
Assuming that we are keeping both, is there any code that can be shared between the two, or are there tests that could cover both?  I didn't look very closely, but just checking that you have tried to avoid code duplication as much as possible.

LongCatIsLooong replied:
The reason for keeping both is the API of `CupertinoSlidingSegmentedControl` is drastically different from `CupertinoSegmentedControl`. It's probably too big of a breaking change (and breaks every existing `CupertinoSegmentedControl`) to introduce in one go. Introducing it as a new widget will allow people to migrate as they feel comfortable. But yeah we will have to maintain both widgets with duplicated code for the time being now.

justinmc replied:
Ah I see, makes sense.  Is it worth adding a `@deprecated` flag to CupertinoSegmentedControl now?

LongCatIsLooong replied:
IMO for now it doesn't hurt for them to coexist, as the comment you mentioned requested.

justinmc reviewed (commented):
Thanks for tackling this!

There are a few questions in my comments below that I wanted to clear up, and a few nitpicks, otherwise looks good.

Does this support dark mode by the way?

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:
```diff
@@ -0,0 +1,1037 @@
+/// [Map] of [children]. The type T is the type of the keys used
+/// to identify each widget and determine which widget is selected. As
+/// required by the [Map] class, keys must be of consistent types
+/// and must be comparable. The ordering of the keys will determine the order
```
Actually I copied these from `CupertinoSegmentedControl` :sweat_smile: . The child widgets will be visually placed in the same order as they appear in the `Map`, if the `Map` is unordered then the children may appear in a different order every time the widget rebuilds.

justinmc replied:
Maybe add a disclaimer like I see elsewhere: `The [children] argument must be an ordered [Map] such as a [LinkedHashMap]`.  I feel like it will be easier for people to end up with the order changing on them without understanding why.

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:
```diff
@@ -0,0 +1,1037 @@
+  /// Creates an iOS-style segmented control bar.
+  ///
+  /// The [children] and [controller] arguments must not be null. The [children]
+  /// argument must be an ordered [Map] such as a [LinkedHashMap]. Further, the
```
You mean to mention `such as a [LinkedHashMap]`?

justinmc replied:
Yeah the part about `The [children] argument must be an ordered [Map] such as a [LinkedHashMap].`.  I'm responding on that comment above (https://github.com/flutter/flutter/pull/42775#discussion_r335096179).

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:
```diff
@@ -0,0 +1,1037 @@
+      duration: _kOpacityAnimationDuration,
+      vsync: this,
+    )..addListener(() {
+      setState(() {
```
The number of `AnimationController`s is proportional to the number of segments. I thought each `AnimatedBuilder` can only be bound to one `Animation`?

justinmc replied:
True, I'm not aware of a way to do that with AnimatedBuilder.

LongCatIsLooong replied:
But the list of child `Listenable`s seems to be immutable once the `_MergingListenable` is constructed using `Listenable.merge`. In my case I need to add new `Listenable`s and remove existing `Listenable`s from the list in `didUpdateWidget`, because developers may want to add/remove segments.

HansMuller replied:
If you build the Listenable with Listenable.merge() when you build the AnimatedBuilder itself, there's no getting out of sync?

LongCatIsLooong committed:
update comments

njovy commented:
```dart
 @override
  Widget build(BuildContext context) {
    super.build(context);
    return CupertinoPageScaffold(
      child: CustomScrollView(slivers: <Widget>[
        CupertinoSliverNavigationBar(
          largeTitle: Text("Test"),
          middle: CupertinoSlidingSegmentedControl<int>(
            children: {
              0: Padding(
                padding: EdgeInsets.symmetric(horizontal: 24, vertical: 6),
                child: Text(
                  "Test1",
                  style: TextStyle(fontSize: 14),
                ),
              ),
              1: Padding(
                padding: EdgeInsets.symmetric(horizontal: 24, vertical: 6),
                child: Text("Test2", style: TextStyle(fontSize: 14)),
              ),
              2: Padding(
                padding: EdgeInsets.symmetric(horizontal: 24, vertical: 6),
                child: Text("Test3", style: TextStyle(fontSize: 14)),
              ),
            },
            controller: ValueNotifier<int>(0),
          ),
        ),
        SliverFillRemaining(
          child: Container(),
        )
      ]),
    );
  }
```
onTap is not called if I use this widget with a Sliver widget inside CustomScrollView
Above is a sample code that scrolling works but tapping.
Thank you for this rapid update!

LongCatIsLooong committed:
Merge remote-tracking branch 'upstream/master' into segmented-control

justinmc commented on packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:
```diff
@@ -0,0 +1,1037 @@
+  /// child widget, or tap on a different child widget. When it is changed programmatically,
+  /// all sliding animations will play as if the new selected child widget was
+  /// tapped on.
+  final ValueNotifier<T> controller;
```
Discussed offline.  I missed that this is passed in by the user at the top level.  Looks good as-is.

justinmc reviewed (approved):
LGTM assuming @njovy's use case works.

Sorry if I spam you with mismatched comments and reviews, I'm failing at using Github right now...

LongCatIsLooong committed:
update gesture handling

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:
```diff
@@ -0,0 +1,1037 @@
+/// widget will appear to have the same dimensions.
+///
+/// A segmented control may optionally be created with custom colors. The
+/// [thumbColor], [backgroundColor] arguments can be used to override the segmented
```
I remember I had the same question and I vaguely remember Hans' answer was themes are complicated so we want to make a careful decision. @HansMuller did I just put words in your mouth?

LongCatIsLooong committed:
make stuff private

LongCatIsLooong commented:
@njovy Thank you so much for trying out the PR and catching the bug! I've updated the gesture handling so it works (seemingly) correctly in conjunction with a `VerticalDragGestureRecognizer` now. 

@justinmc thanks for the review! Updated per your comments.

njovy commented:
@LongCatIsLooong It works like a charm with the latest commit. Thank you so much!

HansMuller commented on packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:
```diff
@@ -0,0 +1,1063 @@
+          borderRadius: const BorderRadius.all(Radius.circular(_kCornerRadius)),
+          color: CupertinoDynamicColor.resolve(widget.backgroundColor, context),
+        ),
+        child: UnconstrainedBox(constrainedAxis: Axis.horizontal, child: box),
```
It's a bit difficult to understand why we need another UnconstrainedBox here. Could _SegmentedControlRenderWidget's renderer  just loosen its horizontal constraints?

HansMuller commented on packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:
```diff
@@ -0,0 +1,1063 @@
+  @override
+  RenderObject createRenderObject(BuildContext context) {
+    return _RenderSegmentedControl<T>(
+      selectedIndex: selectedIndex,
```
On the face of it, it seems odd to pass along selectedIndex and the widget state (which could provide a selectedIndex getter) but not update selectedIndex in updateRenderObject.

LongCatIsLooong replied:
`selectedIndex` is conditionally updated in `updateRenderObject`. If the user is currently dragging the thumb around attempting to update `selectedIndex` programmatically will have no effect.
```
      ..guardedSetHighlightedIndex(selectedIndex);
```

HansMuller commented on packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:
```diff
@@ -0,0 +1,1063 @@
+  double currentThumbScale = 1;
+
+  Offset _localDragOffset;
+  // The drag gesture startd on a selected segment.
```
startd => started

HansMuller commented on packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:
```diff
@@ -0,0 +1,1063 @@
+  @override
+  double computeMinIntrinsicWidth(double height) {
+    RenderBox child = firstChild;
+    double minWidth = 0.0;
```
Best to be consistent about doubles; so 0 here and below

HansMuller replied:
Really maxMinChildWidth or widestMinChildWidth

HansMuller commented on packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:
```diff
@@ -0,0 +1,1063 @@
+    double minWidth = 0.0;
+    while (child != null) {
+      final _SegmentedControlContainerBoxParentData childParentData = child.parentData;
+      final double childWidth = child.getMinIntrinsicWidth(height) + 2 * _kSegmentMinPadding;
```
Could just factor the `_kSegmentMinPadding` in at the end? Like:
```dart
return childCount * (widestMinChildWidth + 2 * _kSegmentMinPadding) + totalSeparatorWidth;
```

HansMuller commented on packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:
```diff
@@ -0,0 +1,1063 @@
+      maxWidth = math.max(maxWidth, childWidth);
+      child = childParentData.nextSibling;
+    }
+    return maxWidth * childCount + totalSeparatorWidth;
```
Similar comments about var name and etc as above.

HansMuller commented on packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:
```diff
@@ -0,0 +1,1063 @@
+
+  @override
+  double computeMinIntrinsicHeight(double width) {
+    RenderBox child = firstChild;
```
Could this be:
```dart
return getChildrenAsList.fold<double>(0, (double minHeight, RenderBox child) {
  return math.max(minHeight, child.getMinIntrinsicHeight(width));
});
```

Here and for computeMaxIntrinsicHeight().

LongCatIsLooong replied:
`getChildrenAsList`'s documentation says
>   /// This function is useful when you need random-access to the children of
  /// this render object. If you're accessing the children in order, consider
  /// walking the child list directly.

I think the imperative way is slightly more efficient?

HansMuller replied:
I don't think the efficiency aspect will matter,  but there's no harm in leaving this code as it is. It's pretty easy to read.

HansMuller commented on packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:
```diff
@@ -0,0 +1,1063 @@
+    if (highlightedIndex != null) {
+      if (_childAnimations == null) {
+        _childAnimations = <RenderBox, _ChildAnimationManifest> { };
+        for (int i = 0; i < childCount - 1; i++) {
```
Please use `i += 1`, here and elsewhere.

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:
```diff
@@ -0,0 +1,1063 @@
+
+// The spring animation used when the thumb changes its rect.
+final SpringSimulation _kThumbSpringAnimationSimulation = SpringSimulation(
+  const SpringDescription(mass: 1, stiffness: 503.551, damping: 44.8799),
```
I got those values from the native `UISegmentedControl`.

LongCatIsLooong committed:
Merge branch 'master' of github.com:flutter/flutter into segmented-control

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:
```diff
@@ -0,0 +1,1063 @@
+    this.padding = _kHorizontalItemPadding,
+    this.backgroundColor = CupertinoColors.tertiarySystemFill,
+  }) : assert(children != null),
+       assert(children.length >= 2),
```
It is documented as:
> Further, the length of the [children] list must be greater than one.

HansMuller replied:
Good point, my mistake.

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:
```diff
@@ -0,0 +1,1063 @@
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+
+    _textDirection = Directionality.of(context);
```
I copied the implementation from `CupertinoSegmentedControl` and didn't change it. Should I remove `_textDirection`?

HansMuller replied:
Yes. It looks like you can remove this entire override.

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:
```diff
@@ -0,0 +1,1063 @@
+  final Map<T, AnimationController> _pressControllers = <T, AnimationController>{};
+  final Tween<double> _pressTween = Tween<double>(begin: 1, end: 0.2);
+
+  List<T> keys;
```
It's used by `indexToKey` which is in turned used by the render object to convert indices (integers) to keys of type `T`.

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:
```diff
@@ -0,0 +1,1063 @@
+      vsync: this,
+    )..addListener(() {
+      setState(() {
+        // The textstyle has changed.
```
The font weight changes if a `Widget` that contains text becomes the currently selected widget.

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:
```diff
@@ -0,0 +1,1063 @@
+  void _didChangeControllerValue() {
+    assert(
+      controller.value == null || widget.children.keys.contains(controller.value),
+      "The controller's value ${controller.value} is not one of the keys in the children map: ${widget.children.keys}",
```
null is a valid value I think? It indicates no segment is currently selected.

HansMuller replied:
Right. What I meant was the the message should say: ... value ${conroller.value} is not null or one of the keys...

LongCatIsLooong committed:
Review

LongCatIsLooong committed:
AnimatedBuilder

LongCatIsLooong committed:
more tests

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/sliding_segmented_control.dart:
```diff
@@ -0,0 +1,1063 @@
+  /// which case no widget will be selected.
+  ///
+  /// The [controller]'s value changes when the user drags the thumb to a different
+  /// child widget, or tap on a different child widget. When it is changed programmatically,
```
Added a new test to verify this works.

LongCatIsLooong committed:
Merge branch 'master' of github.com:flutter/flutter into segmented-control

LongCatIsLooong committed:
Merge remote-tracking branch 'upstream/master' into segmented-control

LongCatIsLooong committed:
update assertion message

HansMuller commented on packages/flutter/test/cupertino/sliding_segmented_control_test.dart:
```diff
@@ -0,0 +1,1091 @@
+
+    final Map<int, Widget> children = <int, Widget>{};
+    children[0] = const SizedBox(
+      height: double.infinity,
```
Maybe include a comment about using height=infinity to ensure that the child consumes all of the available vertical space, like SizedBox.expand() does for both dimensions.

LongCatIsLooong replied:
turns out I didn't need the `SizedBox` since I'm checking the Rect of the `Opacity` widget

HansMuller commented on packages/flutter/test/cupertino/sliding_segmented_control_test.dart:
```diff
@@ -0,0 +1,1091 @@
+    (WidgetTester tester) async {
+      final Map<int, Widget> children = <int, Widget>{};
+      children[0] = const Text('Child 1');
+      children[1] = Container(
```
```dart
children[1] = SizedBox(width: 50, height: 50);
```

HansMuller commented on packages/flutter/test/cupertino/sliding_segmented_control_test.dart:
```diff
@@ -0,0 +1,1091 @@
+      find.byKey(const ValueKey<String>('Segmented Control')),
+    );
+
+    expect(segmentedControl.size.width.isFinite, isTrue);
```
Couldn't we check the actual width of the segmented control here? Could simplify the expected width arithmetic by making the children SizedBox(width: 100) or something similar.

HansMuller commented on packages/flutter/test/cupertino/sliding_segmented_control_test.dart:
```diff
@@ -0,0 +1,1091 @@
+
+    // Vertical drag works for the scroll view.
+    final TestGesture gesture = await tester.startGesture(tester.getCenter(find.text('Child 1')));
+    // The first moveBy doesn't actually move the scrollable. It's there to make
```
Thanks for the explanation

HansMuller commented on packages/flutter/test/cupertino/sliding_segmented_control_test.dart:
```diff
@@ -0,0 +1,1091 @@
+  testWidgets('Insert segment while animation is running', (WidgetTester tester) async {
+    final Map<int, Widget> children = SplayTreeMap<int, Widget>(
+      (int a, int b) => a - b,
+      (dynamic key) => true,
```
Is this parameter needed?

HansMuller reviewed (approved):
LGTM

LongCatIsLooong committed:
review
